gammapy/maps/hpxmap.py

```diff
@@ -117,10 +120,15 @@ def to_hdulist(self, **kwargs):
         extname = kwargs.get('extname', 'SKYMAP')
         # extname_bands = kwargs.get('extname_bands', self.geom.conv.bands_hdu)
         extname_bands = kwargs.get('extname_bands', 'BANDS')
-        hdulist = [fits.PrimaryHDU(), self.make_hdu(**kwargs)]
+        hdu = self.make_hdu(**kwargs)
+        header = hdu.header
+        header.update(self.meta)
```

Comment:
My preferred solution for serializing meta data would be to dump the meta dictionary into a single header keyword (see also my comment to #1250):
```
import json
header['META'] = json.dumps(self.meta)
```
This has several advantages over the FITS keyword approach:
* No potential for collisions with other header keywords.
* No restriction on the length of keywords.  Writing meta data keywords to the FITS header directly will trigger HIERARCH cards for all meta keywords longer than 8 characters.
* Meta data keywords can be case sensitive.  
* Supports meta data with a hierarchical structure via nested dictionaries.

The only real disadvantage I can see is that you lose some readability when manually looking at the header contents (e.g. in fv).

Reply:
@woodmd - Did we ever discuss abandoning FITS completely in favour of something with proper metadata and grouping support, concretely HDF5?

Personally I think it's worth considering, or if we stick with FITS, to use some "hack" like what Matthew proposed to get nicer meta-data support and to get rid of the problem of merging on write / disentangling on write two headers, within FITS.

@JouvinLea @registerrier - Let's discuss tomorrow and get some more opinions from the people involved in CTA metadata at the Gammapy meeting before making a decision here.

Reply:
I'm a big fan of hdf5 myself so I think having some hdf5 support built into gammapy would be a great idea.  However I think abandoning FITS entirely is not realistic in the short-term given the huge amount of existing astro software and tools that rely on this format.  What I could imagine is supporting hdf5 as an alternative output format option (e.g. triggered by a keyword argument to a class `write` method).  At least initially you probably want to write a generic FITS to HDF5 conversion utility so that you don't have to rewrite all of your existing FITS serialization methods.  Maybe a utility like this already exists in another package?

Reply:
Hi @woodmd, @cdeil, maybe one could rely on the external package fits2hdf (http://fits2hdf.readthedocs.io/en/latest/) to generically convert from FITS to HDF5 at the moment? I have never given it a try myself, though, so I don't know if that would suit our needs.

Reply:
Agree that we have to think to maybe support other format but not in this PR...

The thing with this line ```header['META'] = json.dumps(self.meta)``` is that really to acces then it will be a mess but I agree that you can have collisions with other header keywords.

Reply:
I agree that we should defer the discussion of supporting HDF5 (but maybe someone should open an issue thread on this topic).

Regarding the meta data serialization one point I still don't understand is how you will reconstitute the meta dictionary when reading a map back from a FITS HDU.  Did you plan to ingest the whole FITS header into the meta dict?  How will you distinguish between FITS keywords that were part of the meta data and those that are part of the map format spec?  I think this is another place where the `META` keyword approach has some advantages.  To deserialize the meta data you would simply do:
```
self.meta = json.loads(header['META'], object_pairs_hook=OrderedDict)
```

Reply:
I put "HDF5?" on the agenda of the gammapy.maps call next week:
https://github.com/gammapy/gammapy-meetings/tree/master/2018-02-15

I think HDF5 serialisation for our maps would be nice, but for IRFs it would be even nicer (simpler to implement, more likely to be used). In both cases it would help us have a good internal data model and an I/O that can deal with more than one format, the HDF5 that is nicer and something as bad as FITS.

For http://fits2hdf.readthedocs.io/en/latest/ I'm not sure if it's a good idea. If you just want to support HDF5 to get the performance benefits, then it might be the way to go. If you want to also take advantage of better metadata and dataset grouping support in HDF5, then some automatic translation from FITS to HDF5 is not what you want. I didn't really look at it yet, whoever starts to add HDF5 support in Gammapy would have to do some research before putting something ...

I'm +1 to use the JSON META hack for now. It allows us to move on with using META in science analysis, and if someone thinks of a better solution or something emerges in CTA we can switch to that. My understanding is that also in ctapipe they don't want to live with the limitations of FITS, and just store metadata as a JSON serialised string, either in the header or as a binary blob.